FGAP documentation

vramik opened the pull request:
Closes #37245

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

vramik committed:
FGAP documentation

Closes #37245

Signed-off-by: vramik <vramik@redhat.com>

pedroigor commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,227 @@
+
+=== Fine grained admin permissions V2
+
+Sometimes roles like `manage-realm` or `manage-users` are too coarse grain and you want to create
```
As an introduction, I would put something like:

The Fine-Grained Admin Permissions feature allows server administrators to delegate realm management to other administrators, the realm administrators. Different than the regular Role-Based Access Control Mechanism provided through the `Global` and `Realm specific` foles (https://www.keycloak.org/docs/26.1.4/server_admin/#master-realm-access-control), this feature provides a more fine-grained control over how realm resources can be accessed and managed based on a well-defined set of operations that can be performed on them.

By relying on a Policy-Based Access Control, server administrators can define permissions to realm resources such as users, groups, and clients, using different policy types, or access control methods, so that a realm administrator is limited to access a subset of realm resources and their operations.

### Understanding the Realm Resource Types

In this section, I would put what you already have about each resource type and the permissions available for each of them.

### Enabling Fine-Grained Admin Permissions to a realm

### Managing Permissions

You already have some text for this section.

#### Defining permissions for viewing realm resources

#### Searching Permissions

You already have some text for this section.

#### Resolving conflicting permissions

You already have some text for this section.

### Evaluating Permissions

Explain the Evaluate tab

### Accessing a realm as a realm administrator
#### Granting the administrative roles to a realm administrator

### Understanding some common use cases

pedroigor commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,227 @@
+
+===== Client
+
+view::
```
IMO, a table is better to list all the scopes/permissions that can be set to a resource.

pedroigor commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,240 @@
+
+==== Understanding the Realm Resource Types
+
+Administrators can manage access to the following resource types:
```
I'm missing understanding what a realm resource type is:

In a realm, you can manage different types of resources such as users, groups, clients, client scopes, roles, and etc.  As a realm administrator, you are constantly managing these resources to manage your identities and how they authenticate and authorize to a realm and your applications.

Each realm resource supports a well-defined set of management operations, or actions, that can be performed on them, such as `view`, `manage`, and resource-specific operations such as `view-members` if you take groups as an example.

This feature provides the necessary mechanisms to enforce access to a well-defined set of realm resources, limited to:

* `Users`
* `Clients`
* `Groups`
* `Roles`

When managing permissions for a specific realm resource type, you can choose to do it for all resources of a given resource type, such as `Users`, or for a specific realm resource, such as a specific user in the realm.

#### User Realm Resources

The `Users` realm resource type represents the users in a realm. You can manage permissions for  users based on the following set of management operations:

| Operation  | Description |
| ------------- | ------------- |
| `view`  | Defines if a realm administrator can view users.  |
| `manage`  | Defines if a realm administrator can manage users.  |

The same pattern can be used to describe the other resource types ....

vramik commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,373 @@
+  realm-specific admin API and admins defined within those realms. You cannot define cross-realm fine grained permissions.
+* Fine grained admin permissions are used to grant additional permissions. You cannot override the default behavior of the 
+  built-in admin roles.
+* Fine-grained admin permissions V2 do not support federated resources. This is a known limitation under consideration 
+  for future improvement.
```
@pedroigor do you think it is sufficient to mention it here withing "key points"? Or would you suggest some changes?

pedroigor replied:
Works for me to highlight it here. Thanks.

pedroigor commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+  ** Administrators must now explicitly assign each required permission.
+  ** Example: To both `view` and `manage` a resource, both `view` and `manage` scopes for a permissions must be assigned separately.
+* Permission model changes:
+  ** The `user-impersonated` user permission has been **removed** and replaced with the new **`impersonate-members`** scope for groups.
```
@vramik Can you keep in docs only what we have today? Documentation for new scopes should be done by their corresponding PRs.

pedroigor commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+  ** Example: To both `view` and `manage` a resource, both `view` and `manage` scopes for a permissions must be assigned separately.
+* Permission model changes:
+  ** The `user-impersonated` user permission has been **removed** and replaced with the new **`impersonate-members`** scope for groups.
+  ** The `configure` client permission has been removed. More fine-grained scopes are planned for future improvements.
```
@vramik Can you keep in docs only what we have today? Documentation for updating scopes should be done by their corresponding PRs.

pedroigor commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain.adoc:
```diff
@@ -1,10 +1,9 @@
-:tech_feature_name: Fine Grain Admin Permissions
-:tech_feature_id: admin-fine-grained-authz
-include::../templates/techpreview.adoc[]
+IMPORTANT: Fine-grained Admin Permissions V1 have been replaced by a new version. Version 1 of the feature is still marked 
```
Makes sense for me too.

vramik replied:
Done

pedroigor commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,19 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+
+* *Centralized Admin Console Management* - New `Permissions` section was introduced to allow management from a single place without having to navigate to different places in the administration console. 
+* *Resource-Specific and Global Permissions* – Permissions can be defined for individual resources (e.g., specific users or groups),  or entire resource types (e.g., all users or all groups).
+* *Explicit Operation Scoping* – Permissions are now independent, removing hidden dependencies between operations. Admins must assign each scope explicitly, making it easier to see what is granted without needing prior knowledge of implicit relationships.
```
I would add:

* Improved manageability experience so that administrators can more easily search and evaluate permissions when building an permission model for realm resources

Perhaps we should also highlight that v2 brings the same level of access control over realm resources so people should not expect anything different than what is there in v1. However, with more flexibility to support additional realm resource types and scopes.

pedroigor reviewed (commented):
@vramik Overall, the structure of the document provides a natural flow to understand and use the future by converting the core concepts and how to use it from an administrative PoV as well as when accessing it as a realm administrator.

Except for those last comments, I think we can start asking @andymunro for a review.

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
 
+= Supported Fine-Grained Admin Permissions
+
+This release introduces a new supported version of Fine-Grained Admin Permissions: V2, providing enhanced flexibility and control over administrative access within realms. 
```
```suggestion
This release introduces support for a new version of Fine-Grained Admin Permissions. Version 2 (V2) provides enhanced flexibility and control over administrative access within realms. 
```

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+= Supported Fine-Grained Admin Permissions
+
+This release introduces a new supported version of Fine-Grained Admin Permissions: V2, providing enhanced flexibility and control over administrative access within realms. 
+With this feature, administrators can define permissions for administering users, groups, clients, and roles without relying on broad administrative roles. The initial release 
+offers the same level of access control over realm resources as the previous version, with plans to extend its capabilities in future updates. There are some key points:
```
```suggestion
With this feature, administrators can define permissions for administering users, groups, clients, and roles without relying on broad administrative roles. V2 offers the same level of access control over realm resources as the previous version, with plans to extend its capabilities in future versions. Some key points follow:
```

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+With this feature, administrators can define permissions for administering users, groups, clients, and roles without relying on broad administrative roles. The initial release 
+offers the same level of access control over realm resources as the previous version, with plans to extend its capabilities in future updates. There are some key points:
+
+* *Centralized Admin Console Management* - New `Permissions` section was introduced to allow management from a single place without having to navigate to different places in the administration console. 
```
```suggestion
* *Centralized Admin Console Management* - New *Permissions* section was introduced to allow management from a single place without having to navigate to different places in the Admin Console. 
```

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+
+* *Centralized Admin Console Management* - New `Permissions` section was introduced to allow management from a single place without having to navigate to different places in the administration console. 
+* *Improved manageability* - Administrators can more easily search and evaluate permissions when building a permission model for realm resources.
+* *Resource-Specific and Global Permissions* – Permissions can be defined for individual resources (e.g., specific users or groups),  or entire resource types (e.g., all users or all groups).
```
```suggestion
* *Resource-Specific and Global Permissions* – Permissions can be defined for individual resources (such as, specific users or groups),  or entire resource types (such as., all users or all groups).
```

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+* *Centralized Admin Console Management* - New `Permissions` section was introduced to allow management from a single place without having to navigate to different places in the administration console. 
+* *Improved manageability* - Administrators can more easily search and evaluate permissions when building a permission model for realm resources.
+* *Resource-Specific and Global Permissions* – Permissions can be defined for individual resources (e.g., specific users or groups),  or entire resource types (e.g., all users or all groups).
+* *Explicit Operation Scoping* – Permissions are now independent, removing hidden dependencies between operations. Admins must assign each scope explicitly, making it easier to see what is granted without needing prior knowledge of implicit relationships.
```
```suggestion
* *Explicit Operation Scoping* – Permissions are now independent, removing hidden dependencies between operations. Administrators must assign each scope explicitly, making it easier to see what is granted without needing prior knowledge of implicit relationships.
```

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,22 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+* *Explicit Operation Scoping* – Permissions are now independent, removing hidden dependencies between operations. Admins must assign each scope explicitly, making it easier to see what is granted without needing prior knowledge of implicit relationships.
+* *Per-Realm Enablement* – Fine-Grained Admin Permissions can be enabled on a per-realm basis, allowing greater control over adoption and configuration.
+
+Read link:{adminguide_finegrained_link}[Fine Grained Admin Permissions] to learn more.
```
```suggestion
For more details, see link:{adminguide_finegrained_link}[Fine Grained Admin Permissions].
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain.adoc:
```diff
@@ -1,10 +1,9 @@
-:tech_feature_name: Fine Grain Admin Permissions
-:tech_feature_id: admin-fine-grained-authz
-include::../templates/techpreview.adoc[]
+IMPORTANT: Fine-grained Admin Permissions V1 have been replaced by a <<_fine_grain_permissions, new version>>. 
+Version 1 of the feature is still marked as preview and is available, but it may be deprecated and removed 
+in future. To enable it start the server with `--features=admin-fine-grained-authz:v1`.
```
```suggestion
IMPORTANT: Fine-grained Admin Permissions V1 have been replaced by a <<_fine_grain_permissions, new version>>. 
Version 1 of the feature is still marked as preview and is available, but it may be deprecated and removed 
in future. To enable it, start the server with `--features=admin-fine-grained-authz:v1`.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+
+Due to fundamental changes in the permission model, **automatic migration from V1 to V2 is not available**. To simplify the transition:
+
+* A new `admin-permissions` client has been introduced (it is created automatically upon enabling the capability for the realm), which holds the authorization model for FGAP:V2.
```
```suggestion
* A new `admin-permissions` client is introduced. This client is created when you enable the capability for the realm. The client holds the authorization model for FGAP:V2.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+
+* A new `admin-permissions` client has been introduced (it is created automatically upon enabling the capability for the realm), which holds the authorization model for FGAP:V2.
+* The existing FGAP:V1 authorization model remains unchanged within the `realm-management` client.
+* Administrators must **recreate permissions and policies** using the new model, which can be configured in the updated **`Permissions`** section of the Admin Console.
```
```suggestion
* Administrators must **recreate permissions and policies** using the new model, which can be configured in the updated *Permissions* section of the Admin Console.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+
+==== Key Differences Between FGAP:V1 and FGAP:V2
+
+* Realm-level enablement: 
+  ** FGAP:V2 can be enabled per realm using the new **`Admin Permissions`** switch in **`Realm Settings`**.
+* Centralized management:
+  ** The resource-specific "Permissions" tabs (for users, groups, clients, and roles) have been removed.
+  ** A new `Permissions`section provides centralized management for all administrative permissions from a single place in the Admin Console.
```
```suggestion
* Realm-level enablement: 
  ** FGAP:V2 can be enabled for a realm using the new *Admin Permissions* switch in *Realm Settings*.
* Centralized management:
  ** The resource-specific *Permissions* tabs (for users, groups, clients, and roles) have been removed.
  ** A new *Permissions* section provides centralized management for all administrative permissions from a single place in the Admin Console.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+  ** Administrators must now explicitly assign each required permission.
+  ** Example: To both `view` and `manage` a resource, both `view` and `manage` scopes for a permissions must be assigned separately.
+* Permission model changes:
+  ** The `user-impersonated` user permission has been **removed** and replaced with the new **`impersonate-members`** scope for groups.
```
```suggestion
  ** The *user-impersonated* user permission has been _removed_ and replaced with the new *impersonate-members* scope for groups.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+  ** Example: To both `view` and `manage` a resource, both `view` and `manage` scopes for a permissions must be assigned separately.
+* Permission model changes:
+  ** The `user-impersonated` user permission has been **removed** and replaced with the new **`impersonate-members`** scope for groups.
+  ** The `configure` client permission has been removed. More fine-grained scopes are planned for future improvements.
```
```suggestion
  ** The *configure* client permission has been removed. More fine-grained scopes are planned for future improvements.
```

andymunro replied:
```suggestion
  ** The *configure* client permission has been removed. More fine-grained scopes are planned for future improvements.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,43 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
+* Permission model changes:
+  ** The `user-impersonated` user permission has been **removed** and replaced with the new **`impersonate-members`** scope for groups.
+  ** The `configure` client permission has been removed. More fine-grained scopes are planned for future improvements.
+* Flexible resource scoping:
+  ** Unlike V1, where permissions were granted either to **a single resource** (for `clients`, `groups`, and `roles`) or **all resources** (for `users`), V2 introduces greater flexibility.
+  ** Administrators can now define permissions for:
+    *** A **specific resource**
+    *** A **set of selected resources**
+    *** **All resources** of a given type
+    *** This applies to **all resource types**: `clients`, `users`, `groups`, and `roles`.
```
```suggestion
* Flexible resource scoping:
  ** Unlike V1, where permissions were granted either to *a single resource* (for clients, groups, and roles) or **all resources** (for users), V2 introduces greater flexibility.
  ** Administrators can now define permissions for:
    *** A *specific resource*
    *** A *set of selected resources*
    *** *All resources* of a given type
    *** This applies to *all resource types*: clients, users,groups, and roles.
```

andymunro reviewed (changes requested):
Good write-up, @vramik 
I have made my best attempts to give feedback, but you may need to adjust the suggestions based on what actually appears in the Admin Console.
The formatting rule we use is to user *single asterisks* for anything you see in the Admin Console, such as *Permissions*.   I don't see a need for double asterisks.  I think that is only when the term is followed by a colon, semicolon, or some other formatting element.
Also, you can use an _underscore_ for emphasis.
You don't need to use `monospace` for a word such as user if it is not an element on the screen.

vramik commented:
Thank you @andymunro for the review, appreciate the suggestions, I'll incorporate them. 

Just quick question, since I didn't find any suggestion to the main documentation of V2 [1], I just wanted to double-check if you had a chance to go through it as well. Thanks a lot.

[1] https://github.com/keycloak/keycloak/pull/38422/files#diff-2efff0f40b791cb0a9da2562687105524958a1f80cd5c29c194233b3d9e4bffa

andymunro commented:
Okay @vramik. I missed that part. I will check it out now.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+=== Fine grained admin permissions V2
+
+The Fine-Grained Admin Permissions feature allows server administrators to delegate realm management to other administrators, 
+the realm administrators. Different than the regular Role-Based Access Control Mechanism provided through the Global and 
```
```suggestion
the realm administrators. Different from the regular Role-Based Access Control Mechanism provided through the Global and 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+Here are some key points to understand about fine-grained admin permissions:
+
+* Fine grained admin permissions are only available within <<_per_realm_admin_permissions, dedicated admin consoles>> or 
+  realm-specific admin API and admins defined within those realms. You cannot define cross-realm fine grained permissions.
```
```suggestion
  realm-specific admin APIs and administrators defined within those realms. You cannot define cross-realm fine grained permissions.
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+  realm-specific admin API and admins defined within those realms. You cannot define cross-realm fine grained permissions.
+* Fine grained admin permissions are used to grant additional permissions. You cannot override the default behavior of the 
+  built-in admin roles.
+* Fine-grained admin permissions V2 do not support federated resources. This is a known limitation under consideration 
```
```suggestion
* Fine-grained admin permissions V2 do not support federated resources. This restriction is a known limitation under consideration 
```

andymunro replied:
Always follow this with a noun.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+[NOTE]
+====
+To maximize the effectiveness of fine-grained admin permissions while minimizing the impact of broad admin roles, 
+it is recommended to assign the `query-clients`, `query-groups`, and `query-users` roles (along with roles required 
```
```suggestion
it is recommended to assign the *query-clients*, *query-groups*, and *query-users* roles (along with roles required 
```

andymunro replied:
Use *asterisks* if this appears on the screen.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+==== Understanding the Realm Resource Types
+
+In a realm, you can manage different types of resources such as users, groups, clients, client scopes, roles, and etc. 
```
```suggestion
In a realm, you can manage different types of resources such as users, groups, clients, client scopes, roles, and so on. 
```

andymunro replied:
etc is Latin. Always replace with "and so on"

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+This feature provides the necessary mechanisms to enforce access to a well-defined set of realm resources, limited to:
+
+* `Users`
+* `Groups`
+* `Clients`
+* `Roles`
```
```suggestion
* Users
* Groups
* Clients
* Roles
```

andymunro replied:
You don't need any formatting if it is not actually referring to the screen

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+When managing permissions for a specific realm resource type, you can choose to do it for all resources of a given resource 
+type, such as `Users`, or for a specific realm resource, such as a specific user or set of users in the realm.
+
+===== Users Resource Type
```
Replace ` with * throughout this section.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+  client management privileges.
+====
+
+==== Enabling admin permissions to a realm
```
Replace ` with * everywhere you refer to something on the screen.   You only need ` for what you tell the user to type.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+* Log in to the Admin Console.
+* Navigate to `Realm settings`
+* Enable the `Admin Permissions` switch and click `Save`
```
```suggestion
* Enable *Admin Permissions* and click *Save*.
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+To enable fine-grained admin permissions to a realm, follow these steps:
+
+* Log in to the Admin Console.
+* Navigate to `Realm settings`
```
```suggestion
* Click *Realm settings*.
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+The `Policies` tab allows administrators to define conditions using different access control methods (`policy type`) to determine whether 
+a permission should be granted to an administrator attempting to access a specific resource performing some operation. It also 
+supports basic searching capability, based on policy `name` and its `type`. Fine grained admin permissions were implemented on top of 
```
```suggestion
supports basic searching capability, based on policy `name` and its `type`. Fine-grained admin permissions were implemented on top of 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+===== Searching Permissions
+
+The administration console provides several ways to search for permissions, supporting the following capabilities:
```
```suggestion
The Admin Console provides several ways to search for permissions, supporting the following capabilities:
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+* Search for permissions that contain a specific string in their `Name`
+* Search for permissions of a specific resource type, such as `Users`
+* Search for permissions of a specific resource type that apply to a particular resource (e.g., `Users` resource type for user `myadmin`).
+* Search for permissions of a specific resource type with a given scope (e.g., `Users` resource type permissions with the `manage` scope).
```
```suggestion
* Search for permissions of a specific resource type with a given scope (such as `Users` resource type permissions with the `manage` scope).
```

andymunro replied:
e.g. is Latin. Always use such as or for example.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+* Search for permissions of a specific resource type, such as `Users`
+* Search for permissions of a specific resource type that apply to a particular resource (e.g., `Users` resource type for user `myadmin`).
+* Search for permissions of a specific resource type with a given scope (e.g., `Users` resource type permissions with the `manage` scope).
+* Search for permissions of a specific resource type that apply to a particular resource and have a specific scope (e.g., `Users` resource 
```
```suggestion
* Search for permissions of a specific resource type that apply to a particular resource and have a specific scope (such as `Users` resource 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+image:images/fine-grain-evaluation.png[Fine grained permissions evaluation tab]
+
+In the example above it is visible that user `myadmin`, can `manage` user `user-1`. There is also information what permissions were involved 
```
```suggestion
In the example above it is visible that the user `myadmin`, can `manage` user `user-1`. There is also information what permissions were involved 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+IMPORTANT: Fine-grained admin permissions allow you to set up permissions for individual resources or for the resource type itself (e.g., all users, 
+all groups, etc.). If a permission or permissions related to a specific resource exist, the "all-resource" permission is *NOT* taken into account 
+during evaluation. If no specific permission exists, the fallback is to the "all-resource" permission. This approach helps solve scenarios like 
```
```suggestion
during evaluation. If no specific permission exists, the fallback is to the "all-resource" permission. This approach helps address scenarios like 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+TODO: review following based on the actual behavior
+* Specific user/group permissions take precedence over broader all-resource permissions.
+* If multiple permissions apply to a given user or group (e.g., multiple user-specific permissions, or a permission covering a set of 
```
```suggestion
* If multiple permissions apply to a given user or group (such as multiple user-specific permissions, or a permission covering a set of 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+- `query-groups` – Required to query realm groups.  
+- `query-clients` – Required to query realm clients.  
+
+If an administrator is responsible for multiple resource types (e.g., both users and groups), they must have all corresponding `query-*` 
```
```suggestion
If an administrator is responsible for multiple resource types (such as both users and groups), they must have all corresponding `query-*` 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+- The appropriate `query-*` roles, depending on the resource types they need to administer.
+
+Beyond these foundational roles, **fine-grained admin permissions** can be used to define specific administrative capabilities. While fine-grained 
+permissions allow for more granular control over access, they **cannot override the default behavior of built-in admin roles**.  
```
```suggestion
Beyond these foundational roles, *fine-grained admin permissions* can be used to define specific administrative capabilities. While fine-grained 
permissions allow for more granular control over access, they cannot override the default behavior of built-in admin roles.  
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+* Create new permission by clicking `Create permission` button
+* This time choose `Groups` resource type
+* Fill the name, e.g. `"Disallow managing test-admins"`
```
```suggestion
* Fill in the name, such as `"Disallow managing test-admins"`
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+=====  Allow impersonation of users for members of admin group with specific role assigned
+
+- Create a **User Permission** for specific users (or all users) you want to allow impersonation.
+- Create a **Group Policy** allowing access to members of `test-admins`.
+- Create a **Role Policy** allowing access to users assigned the `impersonation-admin` role.
```
```suggestion
- Create a *User Permission* for specific users (or all users) you want to allow impersonation.
- Create a *Group Policy* allowing access to members of `test-admins`.
- Create a *Role Policy* allowing access to users assigned the `impersonation-admin` role.
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+
+===== Blacklist specific users from being impersonated
+
+- Create a **User Permission** for the specific users you want to prevent from being impersonated.
```
```suggestion
- Create a *User Permission* for the specific users you want to prevent from being impersonated.
```

andymunro replied:
You only need two asterisks in very specific situation, such as when the asterisk is followed by a colon, semi-colon, or special formatting characters.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,374 @@
+===== Blacklist specific users from being impersonated
+
+- Create a **User Permission** for the specific users you want to prevent from being impersonated.
+- Create any **Policy** that evaluates to deny (e.g., a user policy with no users selected).
```
```suggestion
- Create any **Policy** that evaluates to deny (such as a user policy with no users selected).
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain.adoc:
```diff
@@ -1,10 +1,9 @@
-:tech_feature_name: Fine Grain Admin Permissions
-:tech_feature_id: admin-fine-grained-authz
-include::../templates/techpreview.adoc[]
+IMPORTANT: Fine-grained Admin Permissions V1 have been replaced by a <<_fine_grain_permissions, new version>>. 
```
Let's standardize on lower case: fine-grained admin permissions

andymunro reviewed (commented):
Additional comments @vramik 
Let's standardize on lower case: fine-grained admin permissions.
Also you only need one asterisk.  Use asterisks for anything you see on the screen. 
Use ` for what you tell the user to type.

vramik commented:
Thank you @andymunro, I've incorporated you suggestions, also I went through it once more and found out few other places where I replaced ` for * etc according to the guidance. 

I squashed the commits and resolved TODOs. It should be ready for final review.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,362 @@
+
+==== Understanding some common use cases
+
+Let's say an admin wants to allow a group of admins to manage all users in the realm except those, which belongs to the
+administrators group. In our example, we have a realm called `test` and a group called `test-admins`. 
+
+===== Allow manage users by group of admins
```
```suggestion
Let's say an administrator wants to allow a group of administrators to manage all users in the realm except those, which belongs to the
administrators group. In our example, we have a realm called `test` and a group called `test-admins`. 

===== Allow manage users by group of administrators
```

andymunro reviewed (changes requested):
Looking good, @vramik Let's go with fine-grained admin permissions (lower case)
Here admin is acceptable as an adjective, but use administrator as a noun,

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,21 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
 
+= Supported Fine-Grained Admin Permissions
+
+This release introduces support for a new version of Fine-Grained Admin Permissions. Version 2 (V2) provides enhanced flexibility and control over administrative access within realms. 
```
```suggestion
This release introduces support for a new version of fine-grained admin permissions. Version 2 (V2) provides enhanced flexibility and control over administrative access within realms. 
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain.adoc:
```diff
@@ -1,10 +1,9 @@
-:tech_feature_name: Fine Grain Admin Permissions
-:tech_feature_id: admin-fine-grained-authz
-include::../templates/techpreview.adoc[]
+IMPORTANT: fine-grained Admin Permissions V1 have been replaced by a <<_fine_grain_permissions, new version>>. 
```
```suggestion
IMPORTANT: fine-grained admin permissions V1 have been replaced by a <<_fine_grain_permissions, new version>>. 
```

andymunro commented on docs/documentation/topics/templates/document-attributes.adoc:
```diff
@@ -48,6 +48,8 @@
 :adminguide_bruteforce_link: {adminguide_link}#password-guess-brute-force-attacks
 :adminguide_eventlistener_name: Event listener
 :adminguide_eventlistener_link: {adminguide_link}#event-listener
+:adminguide_finegrained_name: Fine Grained Admin Permissions
```
```suggestion
:adminguide_finegrained_name: fine-grained admin permissions
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,42 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
 Currently, it is not supported to request offline tokens or exchange a refresh token when the subject token was issued from an offline session. The recommended approach is to exchange for access tokens instead of
 refresh token when possible.
 
+=== Supported Fine-Grained Admin Permissions
```
```suggestion
=== Fine-grained admin permissions supported
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc:
```diff
@@ -89,6 +89,42 @@ the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification
 
+=== Supported Fine-Grained Admin Permissions
+
+Starting with this release, {project_name} introduces **Fine-Grained Admin Permissions V2**, offering an improved and more flexible authorization model for administrative permissions.
```
```suggestion
Starting with this release, {project_name} introduces *fine-grained admin permissions V2*, offering an improved and more flexible authorization model for administrative permissions.
```

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,362 @@
+
+=== Fine grained admin permissions V2
+
+The Fine-Grained Admin Permissions feature allows server administrators to delegate realm management to other administrators, 
```
```suggestion
With fine-grained admin permissions feature, you can delegate realm management to other administrators, 
```

sguilhen commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,362 @@
+Beyond these foundational roles, *fine-grained admin permissions* can be used to define specific administrative capabilities. While fine-grained 
+permissions allow for more granular control over access, they cannot override the default behavior of built-in admin roles.
+This means that if an additional admin role is assigned to a realm administrator, permission evaluation will be bypassed, and access will be 
+granted.  
```
Not sure I understand this.. if you grant any admin-role to a user, then permission evaluation is skipped irrespective of the role?

vramik replied:
The role should be respective to the action. I was trying to say, that if you assign for instance `manage-clients` role to your "realm-admin", no permissions related to client management are evaluated. 

So even if you have permission saying "do not allow to manage this client", the assigned role takes precedence. 

Would you suggest better wording to describe this behavior?

vramik replied:
For now, I have replaced "additional" by "respective".

andymunro commented on docs/documentation/topics/templates/document-attributes.adoc:
```diff
@@ -48,6 +48,8 @@
 :adminguide_eventlistener_name: Event listener
 :adminguide_eventlistener_link: {adminguide_link}#event-listener
+:adminguide_finegrained_name: Fine Grained Admin Permissions
+:adminguide_finegrained_link: {adminguide_link}#_fine_grain_permissions
```
Yes, that makes sense, @sguilhen

andymunro commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -8,6 +8,21 @@ For more details, see the link:{securing_apps_token_exchange_link}#_standard-tok
+
+* *Centralized Admin Console Management* - New *Permissions* section was introduced to allow management from a single place without having to navigate to different places in the Admin Console. 
+* *Improved manageability* - Administrators can more easily search and evaluate permissions when building a permission model for realm resources.
+* *Resource-Specific and Global Permissions* – Permissions can be defined for individual resources (such as, specific users or groups),  or entire resource types (such as., all users or all groups).
```
Yes, such as does not need a comma.  If you use for example, a comma is okay.

andymunro commented on docs/documentation/server_admin/topics/admin-console-permissions/fine-grain-v2.adoc:
```diff
@@ -0,0 +1,362 @@
+
+==== Enabling admin permissions to a realm
+
+To enable fine-grained admin permissions to a realm, follow these steps:
```
Good, @sguilhen I say "in a realm"

pedroigor commented:
@andymunro Are you suggestions addressed? Wondering if we can consider it as done from your PoV. We will be making updates to this doc as we move with some remaining tasks too.

andymunro reviewed (approved):
Thanks, @vramik Looks good to me.

pedroigor commented:
Great work @vramik.
